clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -379,6 +382,45 @@ inlineCleanup (TransformContext is0 _) (Letrec binds body) = do
 inlineCleanup _ e = return e
 {-# SCC inlineCleanup #-}
 
+-- | Takes a binding and collapses its term if it is a noop
+inlineIdentity :: HasCallStack => NormRewrite
+inlineIdentity _ (Letrec binds body) = do
+  (untouched,collapsed) <- partitionEithers <$> mapM identityAnalysis binds
+  if null collapsed
+    then return (Letrec binds body)
+    else changed (Letrec (untouched ++ collapsed) body)
+inlineIdentity _ e = return e
+
+identityAnalysis :: HasCallStack => (Id,Term) -> NormalizeSession (Either (Id,Term) (Id,Term))
+identityAnalysis old@(iD,collectArgs -> (term,partitionEithers -> (termArgs,typeArgs))) =
+  runIdentityAnalysis $ do
+    Prim primInfo <- return term
+    identity      <- getIdentity primInfo
+    let resTy = polyResTyView $ primType primInfo
+    return (iD,primUCo `App` identity `mkTyApps` (resTy:typeArgs))
```

Comment:
With the corrected type for `unsafeCoerce#` this should be the type applications followed by the term application. We have a function `mkApps :: Term -> [Either Term Type] -> Term` which could be used here, i.e.

```haskell
return (iD, mkApps primUnsafeCoerce [Right a, Right b, Left x])
```